Added a simple 2x5 Keypad with 4 layers

jmcameron opened the pull request:
##Description

This Keypad has 2 rows x 5 columns of keys.   It has the top/default layer that has a few handy navigation keys as well as one dedicated key to cycle through the layers.   The second layer has some media keys.  The third and fourth layers are currently undefined.   The keypad also includes three LEDs that show which layer is active (no lit LEDs means the default layer).

## Types of Changes
This is a new keyboard layout.

drashna commented on keyboards/2x5keypad/2x5keypad.h:
```diff
@@ -0,0 +1,23 @@
+#ifndef KB_H
```
If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`? 

It's simpler, and less prone to user error, so if you could make this switch? 

```suggestion
#pragma once
```

drashna commented on keyboards/2x5keypad/2x5keypad.h:
```diff
@@ -0,0 +1,23 @@
+#ifndef KB_H
+#define KB_H
```
```suggestion
```

drashna commented on keyboards/2x5keypad/2x5keypad.h:
```diff
@@ -0,0 +1,23 @@
+void keyboard_pre_init_user(void);
+
+
+#endif
```
```suggestion
```

jmcameron replied:
Done

drashna commented on keyboards/2x5keypad/2x5keypad.h:
```diff
@@ -0,0 +1,23 @@
+#define GREEN_LED B6
+
+
+#define KEYMAP( \
```
We've switched from using the "keymap" working, to "layout"

```suggestion
#define LAYOUT( \
```

drashna commented on keyboards/2x5keypad/config.h:
```diff
@@ -0,0 +1,58 @@
+#ifndef CONFIG_H
```
If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`? 

It's simpler, and less prone to user error, so if you could make this switch? 

```suggestion
#pragma once
```

drashna commented on keyboards/2x5keypad/config.h:
```diff
@@ -0,0 +1,58 @@
+#ifndef CONFIG_H
+#define CONFIG_H
```
```suggestion
```

drashna commented on keyboards/2x5keypad/config.h:
```diff
@@ -0,0 +1,58 @@
+#define RGBLIGHT_VAL_STEP 8
+#endif
+
+#endif
```
```suggestion
```

drashna commented on keyboards/2x5keypad/config.h:
```diff
@@ -0,0 +1,58 @@
+
+/* key combination for command */
+#define IS_COMMAND() ( \
+    keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
```
We've changed how this is called, actually: 

```suggestion
    get_mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
```

And actually, since this using the default value, you could remove the entire `IS_COMMAND` block

drashna commented on keyboards/2x5keypad/config.h:
```diff
@@ -0,0 +1,58 @@
+#endif
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCING_DELAY 5
```
This should be: 

```suggestion
#define DEBOUNCE 5
```

drashna commented on keyboards/2x5keypad/config.h:
```diff
@@ -0,0 +1,58 @@
+)
+
+/* prevent stuck modifiers */
+#define PREVENT_STUCK_MODIFIERS
```
This is actually the default now, and doesn't do anything. 

```suggestion
```

drashna commented on keyboards/2x5keypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,72 @@
+#include "2x5keypad.h"
```
```suggestion
#include QMK_KEYBOARD_H
```

drashna commented on keyboards/2x5keypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,72 @@
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [NORMAL_LAYER]=
+	KEYMAP(
```
```suggestion
	LAYOUT(
```

drashna commented on keyboards/2x5keypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,72 @@
+		WIN_LOCK, KC_MUTE, KC_LEFT, KC_DOWN, KC_RGHT),
+
+    [MEDIA_LAYER]=
+	KEYMAP(
```
```suggestion
	LAYOUT(
```

drashna commented on keyboards/2x5keypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,72 @@
+		KC_TRNS,  KC_TRNS, KC_MSTP, KC_MPLY, KC_VOLD),
+
+    [TBD_LAYER2]=
+	KEYMAP(
```
```suggestion
	LAYOUT(
```

drashna commented on keyboards/2x5keypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,72 @@
+		KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS),
+
+    [TBD_LAYER3]=
+	KEYMAP(
```
```suggestion
	LAYOUT(
```

drashna commented on keyboards/2x5keypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,72 @@
+		KC_TRNS,  KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS)
+};
+
+const macro_t *action_get_macro(keyrecord_t *record, uint8_t id, uint8_t opt) {
```
This has been deprecated and should be removed.

drashna commented on keyboards/2x5keypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,72 @@
+	return true;
+}
+
+uint32_t layer_state_set_user(uint32_t state)
```
We actually use a typedef here, now, so you can reduce the number of layers used. 

```suggestion
layer_state_t layer_state_set_user(layer_state_t state)
```

drashna commented on keyboards/2x5keypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,72 @@
+
+uint32_t layer_state_set_user(uint32_t state)
+{
+    switch (biton32(state))
```
Also, it may be simpler here to turn off ALL of the LEDs, and then only turn on the ones that you need.  

Functionally, it would be the same, but the code would look cleaner, I think.

drashna commented on keyboards/2x5keypad/rules.mk:
```diff
@@ -0,0 +1,56 @@
+# MCU name
+MCU = atmega32u4
```
Actually, everything between this and bootmagic could be removed, as we've recently implemented defaults for many of the MCUs. 

In fact, you could just set this entire file to be: 

```
MCU = atmega32u4
BOOTLOADER = atmel-dfu

BOOTMAGIC_ENABLE = yes	# Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes	# Mouse keys
EXTRAKEY_ENABLE = yes	# Audio control and System control
CONSOLE_ENABLE= no	# Console for debug
COMMAND_ENABLE = no    # Commands for debug and configuration
SLEEP_LED_ENABLE = no  # Breathing sleep LED during USB suspend
NKRO_ENABLE = yes	# USB Nkey Rollover -
BACKLIGHT_ENABLE = no  # Enable keyboard backlight functionality
AUDIO_ENABLE = no
RGBLIGHT_ENABLE = no
```

drashna commented on keyboards/2x5keypad/rules.mk:
```diff
@@ -0,0 +1,56 @@
+
+
+# Boot Section Size in *bytes*
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
Do you know what Boot loader is being used?  

If so, could you specify that rather than the size?  Eg: 
```make
# Bootloader selection
#   Teensy       halfkay
#   Pro Micro    caterina
#   Atmel DFU    atmel-dfu
#   LUFA DFU     lufa-dfu
#   QMK DFU      qmk-dfu
#   atmega32a    bootloadHID
BOOTLOADER = atmel-dfu
```

drashna commented:
Also, this keyboard is lacking a readme file?   If you could, please add one.  You can use the template one, as a base:
https://github.com/qmk/qmk_firmware/blob/master/quantum/template/avr/readme.md


Additionally, is this just a personal keyboard?  Eg, not mass produced, or ran as a group buy? 
If so, could you move it to the `keyboards/handwired` folder, as that is meant more for small runs and personal projects? 


Otherwise, nice macropad!

jmcameron commented:
@drashna Thanks for the suggestions.  I've incorporated all of them (I think).  I also recompiled and reflashed to verify it all works.  Also moved the directory to under hardwired/.  And added a readme file with a picture.

jmcameron commented:
Oops. Sorry about not doing this on a development branch.  Next time!

fauxpark commented on keyboards/handwired/2x5keypad/readme.md:
```diff
@@ -0,0 +1,26 @@
+  * Chassis is three layers of wood
+  * Handwired
+  * Uses a Pro Micro
+  * Includes a reset switch accsible by a hole on the bottom
```
```suggestion
  * Includes a reset switch accessible by a hole on the bottom
```

fauxpark commented:
Could you also please host the image somewhere like imgur instead?

fauxpark commented on keyboards/handwired/2x5keypad/keymaps/default/keymap.h:
```diff
@@ -0,0 +1,16 @@
+};
+
+
+uint32_t layer_state_set_user(uint32_t state);
```
This can be removed; QMK already declares this function.
The rest of this file should go in keymap.c instead - there is no need for the header.

jmcameron replied:
Will delete this (actually will delete keymap.h file).

fauxpark commented on keyboards/handwired/2x5keypad/2x5keypad.h:
```diff
@@ -0,0 +1,20 @@
+}
+
+
+void keyboard_pre_init_user(void);
```
```suggestion
```

fauxpark commented on keyboards/handwired/2x5keypad/2x5keypad.c:
```diff
@@ -0,0 +1,8 @@
+#include "2x5keypad.h"
+
+void keyboard_pre_init_user(void)
```
```suggestion
void matrix_init_kb(void)
```

jmcameron replied:
Ok, I don't understand this suggestion.  Should I replace keyboard_pre_init_user() with matrix_init_kb() or should my keyboard_pre_init_user() call matrix_init_kb() in its body?

fauxpark replied:
`keyboard_pre_init_user()` should be renamed to `matrix_init_kb()` and call `matrix_init_user()` before setting up the LEDs. The core implementation of this function simply calls `matrix_init_user()`, so we need to do the same thing in the overriding function so that it is still called for keymaps.

fauxpark commented on keyboards/handwired/2x5keypad/2x5keypad.h:
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include "quantum.h"
+#include "led.h"
```
```suggestion
```

Just this one last thing (sorry to keep re-reviewing), and it looks good 👍

jmcameron commented:
Thanks for the suggestions.  Could you suggest a keyboard that is cleanly
coded with up-to-date APIs, etc, that I can use as a model when I do
something more complex with this on or do a full keyboard?

Thanks

-Jonathan

On Sun, Sep 8, 2019 at 6:42 PM fauxpark <notifications@github.com> wrote:

> *@fauxpark* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qmk/qmk_firmware/pull/6699?email_source=notifications&email_token=AAAPFQIGYIYZ5EC65WTXDUTQIWS2BA5CNFSM4IUR5FC2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCEAJOJA#pullrequestreview-285251364>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAPFQPDQ3AOEMW3ROLHJMTQIWS2BANCNFSM4IUR5FCQ>
> .
>

fauxpark commented:
I use the [Wasdat](https://github.com/qmk/qmk_firmware/tree/master/keyboards/maartenwut/wasdat) and [GH60 Satan](https://github.com/qmk/qmk_firmware/tree/master/keyboards/gh60/satan), so their configurations are pretty well up to date :)

jmcameron commented:
Thanks

On Sun, Sep 8, 2019 at 7:31 PM fauxpark <notifications@github.com> wrote:

> I use the Wasdat
> <https://github.com/qmk/qmk_firmware/tree/master/keyboards/maartenwut/wasdat>
> and GH60 Satan
> <https://github.com/qmk/qmk_firmware/tree/master/keyboards/gh60/satan>,
> so their configurations are pretty well up to date :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qmk/qmk_firmware/pull/6699?email_source=notifications&email_token=AAAPFQJTIGN4B32KGCCY7ELQIWYRTA5CNFSM4IUR5FC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6GCITY#issuecomment-529278031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAPFQPJQEDSY6OGZBEPDLDQIWYRTANCNFSM4IUR5FCQ>
> .
>

drashna commented on keyboards/handwired/2x5keypad/config.h:
```diff
@@ -0,0 +1,57 @@
+*/
+
+/* prevent stuck modifiers */
+#define PREVENT_STUCK_MODIFIERS
```
This should be removed, since it's the default behavior now, and doesn't actually do anything. 

```suggestion
```

drashna commented:
Awesome, thanks!
